src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c

```diff
@@ -642,6 +643,40 @@ bool dcd_edpt_open (uint8_t rhport, tusb_desc_endpoint_t const * p_endpoint_desc
   return true;
 }
 
+/**
+ * Close an endpoint.
+ * 
+ * This function should be called with interrupts enabled, though 
```

Comment:
there is no need to force the interrupt disabled or eanbled. If the dcd need an interrupt disabled, it could call dcd_init_disable() within this function.

Reply:
dcd_int_enable doesn't keep track of how many times it's called. The queue manipulation code needs interrupts disabled to run properly. If the interrupt were disabled when the function was called, then it would be reenabled prematurely.

FreeRTOS uses a counter to keep track of the nesting depth of disable calls (like a stack) so nested calls can be OK.

Reply:
I mean the dcd API doesn't impose any requirement of interrupt state, if need, dcd just get the current interrupt state via NVIC, disable, and enable it if needed. The stack won't impose that requirement, unless we run into issue that need it.

Reply:
I could be wrong, thing are not really clear from my point of view now, maybe we could figure this out in the process.

Reply:
Yeah, but NVIC is ARM-specific (i think?).  There needs to be more BSP or DCD code to do refcounts. It was just easier to impose a limitation on calling the function, for now.

Reply:
Maybe create usbd_int_enterCritical or something to do refcount, and that calls the dcd functions. But it then creates a whole series of headaches since you don't want to reenable interrupts, if already in an interrupt handler.

Reply:
it is inside dcd, it is specific to stm32 already. We could impose it when it is needed across ports. Since stm32 fsdev work with interrupt enabled or disabled, I don't see a reason to impose any limit just now. But it may just be me not seeing the whole picture for now. We can of course address it in the process.

Reply:
Oh, it's way too late here. I was thinking we were talking about usbd.c. My cell phone screen is too small.

You're right that it doesn't matter in this case. Apologies.

I'll fix up this and the other issues in the morning.

Reply:
> Maybe create usbd_int_enterCritical or something to do refcount, and that calls the dcd functions. But it then creates a whole series of headaches since you don't want to reenable interrupts, if already in an interrupt handler.

Then just implement dcd_edpt_close() as it is, I still don't get why you need to eanble/disable interrupt or mutex access. If needed it will be perform by usbd, dcd just implement its part to close the endpoint.

Reply:
It is needed for the queue filtering. If interrupts were enabled, new events could end up in the middle of the queue.

Reply:
@pigrew yeah, it makes sense now, but dcd should not assume usbd will perform any filtering on its queue. We may not end up use that filtering at all. Let's just have dcd close its opened endpoints. The closed endpoints should be best in disabled state. There is Alt interface that has more endpoints than others, and or with complete new configure (Interrupt --> Bulk) etc ..

Reply:
> Oh, it's way too late here. I was thinking we were talking about usbd.c. My cell phone screen is too small.
> 
> You're right that it doesn't matter in this case. Apologies.
> 
> I'll fix up this and the other issues in the morning.

Yeah, I also missed this reply. Sorry for being picky, I only mean to remove the comment on top of the `dcd_edpt_close()` :). Yeah, please take some sleep, it must be late on your time (mine is only midnoon for now :) )

Reply:
I believe this is resolved now?